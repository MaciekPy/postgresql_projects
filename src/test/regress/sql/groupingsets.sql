select a, sum(b) from (values (1,10),(1,20),(2,40)) v(a,b) group by rollup (a);

select a, b, sum(c) from (values (1,1,10),(1,1,11),(1,2,12),(1,2,13),(1,3,14),(2,3,15),(3,3,16),(3,4,17),(4,1,18),(4,1,19)) v(a,b,c) group by rollup (a,b); 

select (select grouping(a,b) from (values (1)) v2(b)) from (values (1)) v1(a) group by a;

select grouping(p), percentile_disc(p) within group (order by x::float8), array_agg(p)
from generate_series(1,5) x,
     (values (0::float8),(0.1),(0.25),(0.4),(0.5),(0.6),(0.75),(0.9),(1)) v(p)
group by rollup (p) order by p;

select a, array_agg(b) from (values (1,10),(1,20),(2,40)) v(a,b) group by rollup (a);

select grouping(a), array_agg(b) from (values (1,10),(1,20),(2,40)) v(a,b) group by rollup (a);

select a, sum(b) from aggtest v(a,b) group by rollup (a);

select grouping(a), sum(b) from aggtest v(a,b) group by rollup (a);

select (select grouping(a,b) from (values (1)) v2(c)) from (values (1,2)) v1(a,b) group by a,b;
