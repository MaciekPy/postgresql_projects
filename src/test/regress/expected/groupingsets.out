select a, b from (values (1,2)) v(a,b) group by rollup (a,b);
 a | b 
---+---
 1 | 2
 1 |  
   |  
(3 rows)

select a, sum(b) from (values (1,10),(1,20),(2,40)) v(a,b) group by rollup (a);
 a | sum 
---+-----
 1 |  30
 2 |  40
   |  70
(3 rows)

select a, b, sum(c) from (values (1,1,10),(1,1,11),(1,2,12),(1,2,13),(1,3,14),(2,3,15),(3,3,16),(3,4,17),(4,1,18),(4,1,19)) v(a,b,c) group by rollup (a,b); 
 a | b | sum 
---+---+-----
 1 | 1 |  21
 1 | 2 |  25
 1 | 3 |  14
 1 |   |  60
 2 | 3 |  15
 2 |   |  15
 3 | 3 |  16
 3 | 4 |  17
 3 |   |  33
 4 | 1 |  37
 4 |   |  37
   |   | 145
(12 rows)

select (select grouping(a,b) from (values (1)) v2(b)) from (values (1)) v1(a) group by a;
ERROR:  Arguments to GROUPING must be grouping expressions of the associated query level
LINE 1: select (select grouping(a,b) from (values (1)) v2(b)) from (...
                                ^
select grouping(p), percentile_disc(p) within group (order by x::float8), array_agg(p)
from generate_series(1,5) x,
     (values (0::float8),(0.1),(0.25),(0.4),(0.5),(0.6),(0.75),(0.9),(1)) v(p)
group by rollup (p) order by p;
 grouping | percentile_disc |                                                                                  array_agg                                                                                  
----------+-----------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
        0 |               1 | {0,0,0,0,0}
        0 |               1 | {0.1,0.1,0.1,0.1,0.1}
        0 |               2 | {0.25,0.25,0.25,0.25,0.25}
        0 |               2 | {0.4,0.4,0.4,0.4,0.4}
        0 |               3 | {0.5,0.5,0.5,0.5,0.5}
        0 |               3 | {0.6,0.6,0.6,0.6,0.6}
        0 |               4 | {0.75,0.75,0.75,0.75,0.75}
        0 |               5 | {0.9,0.9,0.9,0.9,0.9}
        0 |               5 | {1,1,1,1,1}
        1 |               5 | {0,0,0,0,0,0.1,0.1,0.1,0.1,0.1,0.25,0.25,0.25,0.25,0.25,0.4,0.4,0.4,0.4,0.4,0.5,0.5,0.5,0.5,0.5,0.6,0.6,0.6,0.6,0.6,0.75,0.75,0.75,0.75,0.75,0.9,0.9,0.9,0.9,0.9,1,1,1,1,1}
(10 rows)

select a, array_agg(b) from (values (1,10),(1,20),(2,40)) v(a,b) group by rollup (a);
 a | array_agg  
---+------------
 1 | {10,20}
 2 | {40}
   | {10,20,40}
(3 rows)

select grouping(a), array_agg(b) from (values (1,10),(1,20),(2,40)) v(a,b) group by rollup (a);
 grouping | array_agg  
----------+------------
        0 | {10,20}
        0 | {40}
        1 | {10,20,40}
(3 rows)

select a, sum(b) from aggtest v(a,b) group by rollup (a);
  a  |   sum   
-----+---------
   0 | 0.09561
  42 |  324.78
  56 |     7.8
 100 |  99.097
     | 431.773
(5 rows)

select grouping(a), sum(b) from aggtest v(a,b) group by rollup (a);
 grouping |   sum   
----------+---------
        0 | 0.09561
        0 |  324.78
        0 |     7.8
        0 |  99.097
        1 | 431.773
(5 rows)

select (select grouping(a,b) from (values (1)) v2(c)) from (values (1,2)) v1(a,b) group by a,b;
 grouping 
----------
        0
(1 row)

SELECT four, ten, SUM(SUM(four)) OVER (PARTITION BY four), AVG(ten) FROM tenk1
GROUP BY ROLLUP(four, ten) ORDER BY four, ten;
 four | ten |  sum  |          avg           
------+-----+-------+------------------------
    0 |   0 |     0 | 0.00000000000000000000
    0 |   2 |     0 |     2.0000000000000000
    0 |   4 |     0 |     4.0000000000000000
    0 |   6 |     0 |     6.0000000000000000
    0 |   8 |     0 |     8.0000000000000000
    0 |     |     0 |     4.0000000000000000
    1 |   1 |  5000 | 1.00000000000000000000
    1 |   3 |  5000 |     3.0000000000000000
    1 |   5 |  5000 |     5.0000000000000000
    1 |   7 |  5000 |     7.0000000000000000
    1 |   9 |  5000 |     9.0000000000000000
    1 |     |  5000 |     5.0000000000000000
    2 |   0 | 10000 | 0.00000000000000000000
    2 |   2 | 10000 |     2.0000000000000000
    2 |   4 | 10000 |     4.0000000000000000
    2 |   6 | 10000 |     6.0000000000000000
    2 |   8 | 10000 |     8.0000000000000000
    2 |     | 10000 |     4.0000000000000000
    3 |   1 | 15000 | 1.00000000000000000000
    3 |   3 | 15000 |     3.0000000000000000
    3 |   5 | 15000 |     5.0000000000000000
    3 |   7 | 15000 |     7.0000000000000000
    3 |   9 | 15000 |     9.0000000000000000
    3 |     | 15000 |     5.0000000000000000
      |     | 15000 |     4.5000000000000000
(25 rows)

select a, b from (values (1,2),(2,3)) v(a,b) group by grouping sets((a,b),());
 a | b 
---+---
 1 | 2
 2 | 3
   |  
(3 rows)

select a, b from (values (1,2),(2,3)) v(a,b) group by rollup((a,b));
 a | b 
---+---
 1 | 2
 2 | 3
   |  
(3 rows)

select a, b, sum(c) from (values (1,1,10,5),(1,1,11,5),(1,2,12,5),(1,2,13,5),(1,3,14,5),(2,3,15,5),(3,3,16,5),(3,4,17,5),(4,1,18,5),(4,1,19,5)) v(a,b,c,d) group by rollup ((a,b));
 a | b | sum 
---+---+-----
 1 | 1 |  21
 1 | 2 |  25
 1 | 3 |  14
 2 | 3 |  15
 3 | 3 |  16
 3 | 4 |  17
 4 | 1 |  37
   |   | 145
(8 rows)

create temp view tv2(a,b,c,d,e,f,g) as select a[1], a[2], a[3], a[4], a[5], a[6], generate_series(1,3) from (select (array[1,1,1,1,1,1])[1:6-i] || (array[2,2,2,2,2,2])[7-i:6] as a from generate_series(0,6) i) s;
select a,b, sum(g) from tv2 group by grouping sets ((a,b,c),(a,b));
 a | b | sum 
---+---+-----
 1 | 1 |  24
 1 | 1 |   6
 1 | 1 |  30
 1 | 2 |   6
 1 | 2 |   6
 2 | 2 |   6
 2 | 2 |   6
(7 rows)

CREATE TEMP TABLE testgs_emptytable(a int,b int,c int);
SELECT sum(a) FROM testgs_emptytable GROUP BY ROLLUP(a,b);
 sum 
-----
    
(1 row)

SELECT grouping(four), ten FROM tenk1
GROUP BY ROLLUP(four, ten) ORDER BY four, ten;
 grouping | ten 
----------+-----
        0 |   0
        0 |   2
        0 |   4
        0 |   6
        0 |   8
        0 |    
        0 |   1
        0 |   3
        0 |   5
        0 |   7
        0 |   9
        0 |    
        0 |   0
        0 |   2
        0 |   4
        0 |   6
        0 |   8
        0 |    
        0 |   1
        0 |   3
        0 |   5
        0 |   7
        0 |   9
        0 |    
        1 |    
(25 rows)

SELECT grouping(four), ten FROM tenk1
GROUP BY ROLLUP(four, ten) ORDER BY ten;
 grouping | ten 
----------+-----
        0 |   0
        0 |   0
        0 |   1
        0 |   1
        0 |   2
        0 |   2
        0 |   3
        0 |   3
        0 |   4
        0 |   4
        0 |   5
        0 |   5
        0 |   6
        0 |   6
        0 |   7
        0 |   7
        0 |   8
        0 |   8
        0 |   9
        0 |   9
        0 |    
        0 |    
        0 |    
        0 |    
        1 |    
(25 rows)

